### PR TITLE
Omniglot Dataset

### PR DESCRIPTION
**Pull request by activatedgeek**
This is the loader for [Omniglot Dataset](https://github.com/brendenlake/omniglot).

One of the use cases of this dataset is for One-Shot Learning where we sample a pair from the dataset and train a Neural Network to learn the similarity metric between the pair.

P.S.: It is amazing how simple it is to write data loaders!

### CODE REVIEW

**Comment by activatedgeek** [comment]
I am particularly concerned as to how to implement the `__len__` function for the randomized pair. Should I generate a preset number of pairs and just return them on every call to `__getitem__`?

**Comment by activatedgeek** [comment]
I precompute the random set of pairs with an added argument of `pair_count` to the new `OmniglotRandomPair` class constructor. I think this should solve the random pair generation problem and make it deterministic after the object is instantiated.

File: torchvision/datasets/omniglot.py [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+        force_extract (bool, optional): If true, extracts the downloaded zip file irrespective
+            of the existence of an extracted folder with the same name
+    """
+    folder = 'omniglot-py'
+    download_url_prefix = 'https://github.com/brendenlake/omniglot/raw/master/python'
+    zips_md5 = [
+        ['images_background', '68d2efa1b9178cc56df9314c21c6e718'],
+        ['images_evaluation', '6b91aef0f799c5bb55b94e3f2daec811'],
+        # Provision in future
+        # ['images_background_small1', 'e704a628b5459e08445c13499850abc4'],
```
Comment (alykhantejani) [inline torchvision/datasets/omniglot.py]: Please can you remove the commented out bits and open an issue to add these once this is merged?

File: torchvision/datasets/omniglot.py [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+        if download:
+            self.download(force_extract)
+
+        if not self._check_integrity():
+            raise RuntimeError('Dataset not found or corrupted.' +
+                               ' You can use download=True to download it')
+
+        self.target_folder = os.path.join(self.root, self._get_target_folder())
+        self.alphabets_ = list_dir(self.target_folder)
+        self.characters_ = list(
```
Comment (alykhantejani) [inline torchvision/datasets/omniglot.py]: same comment here about trailing underscore

File: torchvision/datasets/omniglot.py [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+            reduce(
```
Comment (alykhantejani) [inline torchvision/datasets/omniglot.py]: nit: the `reduce(lambda...)` can be replaced by `sum([[join(alphabet, character) for c in join(target, alphabet)] for alphabet in _alphabets], [])` i.e. you can use `sum` to do the flatten

File: torchvision/datasets/omniglot.py [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+            )
+        )
+        self.character_images_ = [
+            [
+                tuple([image, idx])
+                for image in list_files(os.path.join(self.target_folder, character), '.png')
+            ]
+            for idx, character in enumerate(self.characters_)
+        ]
+        self.flat_character_images_ = list(reduce(lambda x, y: x + y, self.character_images_))
```
Comment (alykhantejani) [inline torchvision/datasets/omniglot.py]: nit: same comment here, you can use `sum(self.character_images_, [])` instead

File: torchvision/datasets/omniglot.py [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+                        os.path.join(alphabet, character)
+                        for character in list_dir(os.path.join(self.target_folder, alphabet))
+                    ]
+                    for alphabet in self.alphabets_
+                ]
+            )
+        )
+        self.character_images_ = [
+            [
+                tuple([image, idx])
```
Comment (alykhantejani) [inline torchvision/datasets/omniglot.py]: this can just be `(image, idx)`, no need to explicitly call `tuple`

File: torchvision/datasets/omniglot.py [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+            if not os.path.isdir(os.path.join(self.root, filename)) or force_extract is True:
+                print('Extracting downloaded file: ' + os.path.join(self.root, zip_filename))
+                with zipfile.ZipFile(os.path.join(self.root, zip_filename), 'r') as zip_file:
+                    zip_file.extractall(self.root)
+
+    def _get_target_folder(self):
+        return 'images_background' if self.background is True else 'images_evaluation'
+
+
+class OmniglotRandomPair(Omniglot):
```
Comment (alykhantejani) [inline torchvision/datasets/omniglot.py]: Is this a standard dataset? I couldn't see any reference to it on https://github.com/brendenlake/omniglot

If it's a dataset for your use case only, I'd be less inclined to keep it in torchvision. However if it's used widely, then its also fine by me

File: .gitignore [hunk 7]
```diff
@@ -5,4 +5,5 @@ torchvision.egg-info/
 */**/*.pyc
 */**/*~
 *~
-docs/build
\ No newline at end of file
+docs/build
+.idea/
```
[the remark this answers is not in the document]
Comment (activatedgeek) [reply .gitignore]: Sorry about that. Should have already put that in my global gitignore.

File: torchvision/datasets/omniglot.py [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+
+        if download:
+            self.download(force_extract)
+
+        if not self._check_integrity():
+            raise RuntimeError('Dataset not found or corrupted.' +
+                               ' You can use download=True to download it')
+
+        self.target_folder = os.path.join(self.root, self._get_target_folder())
+        self.alphabets_ = list_dir(self.target_folder)
```
[the remark this answers is not in the document]
Comment (activatedgeek) [reply torchvision/datasets/omniglot.py]: Oh! I kind of interpreted the other way, hence the notation.

File: torchvision/datasets/omniglot.py (hunk 3 above)
Comment (activatedgeek) [reply torchvision/datasets/omniglot.py]: Neat! That is more succinct.

File: torchvision/datasets/omniglot.py [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+        return True
+
+    def download(self, force_extract=False):
+        import zipfile
+
+        if self._check_integrity():
+            print('Files already downloaded and verified')
+            return
+
+        for fzip in self.zips_md5:
```
[the remark this answers is not in the document]
Comment (activatedgeek) [reply torchvision/datasets/omniglot.py]: Oh yes, definitely both belonged separately. The `force_extract` was for an adversary scenario I had imagined but perhaps it doesn't make sense. I realize it was not worth the added complexity.

File: torchvision/datasets/omniglot.py (hunk 6 above)
Comment (activatedgeek) [reply torchvision/datasets/omniglot.py]: This dataset is being used for One Shot Learning which has a verification sub-task in the case of Metric Learning. I'll confess that I have a high bias here given that I have only read the papers where the verification task of comparing two images was needed. But I believe, it is valuable to anyone who will be using this closely related task to One Shot Learning.

"Widely used" is something that I cannot completely vouch for.

File: torchvision/datasets/omniglot.py [hunk 10]
```diff
@@ -0,0 +1,195 @@
...
+        filename = self._get_target_folder()
+        zip_filename = filename + '.zip'
+        url = self.download_url_prefix + '/' + zip_filename
+        download_url(url, self.root, zip_filename, self.zips_md5[filename])
+        print('Extracting downloaded file: ' + os.path.join(self.root, zip_filename))
+        with zipfile.ZipFile(os.path.join(self.root, zip_filename), 'r') as zip_file:
+            zip_file.extractall(self.root)
+
+    def _get_target_folder(self):
+        return 'images_background' if self.background is True else 'images_evaluation'
```
Comment (alykhantejani) [inline torchvision/datasets/omniglot.py]: you can remove the `is True` here

File: torchvision/datasets/omniglot.py [hunk 11]
```diff
@@ -0,0 +1,195 @@
...
+        if download:
+            self.download()
+
+        if not self._check_integrity():
+            raise RuntimeError('Dataset not found or corrupted.' +
+                               ' You can use download=True to download it')
+
+        self.target_folder = os.path.join(self.root, self._get_target_folder())
+        self._alphabets = list_dir(self.target_folder)
+        self._characters = sum(
```
Comment (alykhantejani) [inline torchvision/datasets/omniglot.py]: nit: this can be done on one line, which IMO makes it more readable, by choosing less verbose variable names and importing `join` directly i.e.:
```python
from os.path import join
self._characters = sum([[join(a, c) for c in list_dir(join(self.target_folder, a))] for a in self._alphabets], [])
```

File: torchvision/datasets/omniglot.py [hunk 12]
```diff
@@ -0,0 +1,195 @@
...
+    a randomized pair of images with similarity label (0 or 1)
+
+    Args:
+        pair_count (int, optional): The total number of image pairs to generate. Defaults to
+            10000
+    """
+    def __init__(self, root, pair_count=10000, background=True,
+                 transform=None, target_transform=None,
+                 download=False):
+        super(self.__class__, self).__init__(root, background=background,
```
Comment (alykhantejani) [inline torchvision/datasets/omniglot.py]: usually we dont use `self.__class__` but instead do `super(OmniglotRandomPair, self).__init__(...)`

File: torchvision/datasets/omniglot.py [hunk 13]
```diff
@@ -0,0 +1,195 @@
...
+            [
+                [
+                    os.path.join(alphabet, character)
+                    for character in list_dir(os.path.join(self.target_folder, alphabet))
+                ]
+                for alphabet in self._alphabets
+            ],
+            []
+        )
+        self._character_images = [
```
Comment (alykhantejani) [inline torchvision/datasets/omniglot.py]: This doesn't look like it's used in this class, why not just wrap this with the sum and use it everywhere you use `_flat_character_images`. You can probably also make this span a few less lines by bunching the `[[` etc.

File: torchvision/datasets/omniglot.py (hunk 6 above)
Comment (alykhantejani) [reply torchvision/datasets/omniglot.py]: The implementation of this dataset seems a bit complex for what it's trying to achieve. If all you want are random pairs of images and whether they are the same  character or not, then why not just sample on (img, idx) pair from `self._flattened_character_images` and then sample another continuously until they either match or not depending on what `is_match` is. 

There's no need to precompute pairs, just make `__len__` return then length of `self._flattened_character_images` and then when this is used with the pytorch DataLoader, it will stop sampling after `__len__` samples.

File: torchvision/datasets/omniglot.py (hunk 12 above)
Comment (activatedgeek) [reply torchvision/datasets/omniglot.py]: Fair enough. Though any major reason why `self.__class__` would be a bad practice?

File: torchvision/datasets/omniglot.py (hunk 6 above)
Comment (activatedgeek) [reply torchvision/datasets/omniglot.py]: If I keep `__len__` to be equal to the length of `self._flattened_character_images`, that is kind of not really the true length, hence I did not want to conflate those two ideas and put a limitation on the true nature of what I am trying to achieve. (The total possible number of pairs are very large and hence the `pair_count` limit).

I wanted to pre-compute the samples so that every time I use the data loader on the instance, I get the exact same set of pairs. Now that this has come up, I realized that I should also add a `random_seed` argument to make sure it is truly reproducible. Would that make sense?

File: torchvision/datasets/omniglot.py (hunk 13 above)
Comment (activatedgeek) [reply torchvision/datasets/omniglot.py]: I wanted to have an easy way to sample images from a given kind of character id for the randomized pair generation. I discuss this more in the other comment.

**Comment by activatedgeek** [comment]
Hey @alykhantejani, do you mind checking some updates to the randomized pair generation?

**Comment by activatedgeek** [comment]
@alykhantejani Pinging here again. Happy to fix/redo anything that needs more work. I hope you got my earlier comments. I don't seem to find them here though.

**Comment by activatedgeek** [comment]
@alykhantejani Writing the comments on the approach here because I couldn't find it anywhere (weirdly)

> If all you want are random pairs of images and whether they are the same character or not, then why not just sample on (img, idx) pair from `self._flattened_character_images` and then sample another continuously until they either match or not depending on what is_match is.

I believe this way, it is certainly harder to arrive at a collision when `is_match=1`. The way I have maintained the internal data helps me quickly arrive at a pair for both `is_match = {0,1}` because once I flatten the array, I lose information about the character class.

> There's no need to precompute pairs, just make `__len__` return then length of `self._flattened_character_images` and then when this is used with the pytorch DataLoader, it will stop sampling after `__len__` samples.

The problem with this approach would be that I am not staying true to what I wanted to achieve with the class. The total number of pairs combinatorially possible are huge and that is why I introduced a `pair_count` defaulting to 10000 and then set that as the length. It gives the user enough flexibility to sample training pairs as well as setup a One-Shot task. If I limit it to the length of the flattened character list, it gives out the wrong perception and might confuse the user.

**Comment by alykhantejani** [comment]
Hey @activatedgeek, sorry for the response. 

Pinging @fmassa in case he has any opinions/thoughts on the random pair dataset

**Comment by fmassa** [comment]
Hi @activatedgeek , sorry for the delay in replying.

So, my first thought about the `Pair` dataset is the following: we should try not to add sampling logic into the dataset itself, which should only hold information on how to load individual elements, and let the sampling be performed by the `Sampler`/`DataLoader`.
It might look non-trivial in this case, but here is one possibility:

We introduce a new `Dataset` class, in a similar spirit of `ConcatDataset`, that performs `n` samples of the data and returns whatever you want?
Here is a draft implementation (untested):

```python
class MultiDataset(object):
    def __init__(self, dataset, num_outputs=1, transforms=None):
        self.dataset = dataset
        self.num_outputs = num_outputs
        self.transforms = transforms

    def __getitem__(self, idx):
        # here comes the logic to convert a 1d index into a 
        # self.num_output indices, each of size len(self.dataset)
        individual_idx = []
        for i in range(self.num_outputs):
            individual_idx.append(idx % len(self.dataset))
            idx = idx // len(self.dataset)
        
        result = []
        for i in reversed(idx):
            result.append(self.dataset[i])

        if self.transforms is not None:
            result = self.transforms(result)

        return result    

    def __len__(self):
        return len(self.dataset) ** self.num_outputs
```

This way, you generate on-the-fly an arbitrarily large dataset, that can accommodate pairs/triplets/etc of elements of the same dataset. Plus, the logic of how to combine the different targets of each dataset becomes something that the user should do (via the transforms in the `MultiDataset`, each individual dataset has its own `transforms` as well).
And, if you want a biased sampler (say to have a roughly equal number of matches / non-matches) this could be handled by the sampler maybe?

This is just a rough idea, but let me know what you think.

**Comment by activatedgeek** [comment]
@fmassa That is a great idea. I was in fact wondering the same because I recently came across an obvious requirement of similar kind for the ImageNet/Mini-ImageNet datasets as well. It didn't feel right to create custom rules every now and then.

So here is what I will do - the `Omniglot` data loader will return a tuple `(image, character_class_id)` and I will wrap up this PR.

I will take up a `MultiDataset` like utility wrapper in another PR.

**Comment by fmassa** [comment]
Sounds good! Also, giving how generic this dataset is, might be worth considering sending it to pytorch/tnt, but we can discuss that later

**Comment by activatedgeek** [comment]
@fmassa @alykhantejani Can you please verify if everything is in order now?

**Comment by activatedgeek** [comment]
Any updates here?

**Comment by activatedgeek** [comment]
Hi, I am wondering what the hold up is about? Is there something that I could further help with? It has been quite a while that this PR has been up and I would really appreciate if we could make some progress here. (Or closed if it doesn't align well).

 cc @alykhantejani @fmassa

**Comment by activatedgeek** [comment]
@alykhantejani @fmassa Any possibility of this getting merged?

**Comment by fmassa** [comment]
@activatedgeek sorry for the delay, I'll have a look at it today.

**Comment by fmassa** [comment]
This looks very good, thanks! Once the merge conflicts are addressed, I think this can be merged.

**Comment by activatedgeek** [comment]
Hey @alykhantejani @fmassa , can we merge this? I can continue on a discussion on #338 after this.

**Comment by activatedgeek** [comment]
Pinging @fmassa and @alykhantejani here again. Can we please merge this?

**Comment by fmassa** [comment]
Awesome! Thanks a lot @activatedgeek !

**Comment by activatedgeek** [comment]
Hey @fmassa , do you mind taking a look at #338 ?
